runOn:
    -
        # TODO: Update once SERVER-41245 is released
        minServerVersion: "4.3.x"
        topology: ["replicaset, sharded"]

database_name: &database_name "transaction-tests"
collection_name: &collection_name "test"

data: []
tests:
  - description: abortTransaction only retries once after RetryableWriteError from server
    failPoint:
      configureFailPoint: failCommand
      mode: { times: 2 }
      data:
        failCommands: ["abortTransaction"]
        errorLabels: ["RetryableWriteError"]
        errorCode: 112 # WriteConflict, not a retryable error code
    operations:
      - name: startTransaction
        object: session0
      - name: insertOne
        object: collection
        arguments:
        session: session0
        document:
          _id: 1
        result:
          insertedId: 1
      - name: abortTransaction
        object: session0
    expectations:
      - command_started_event:
          command:
            insert: *collection_name
            documents:
              - _id: 1
            ordered: true
            readConcern:
            lsid: session0
            txnNumber:
              $numberLong: "1"
            startTransaction: true
            autocommit: false
            writeConcern:
          command_name: insert
          database_name: *database_name
      - command_started_event:
          command:
            abortTransaction: 1
            lsid: session0
            txnNumber:
            $numberLong: "1"
            startTransaction:
            autocommit: false
            writeConcern:
          command_name: abortTransaction
          database_name: admin
      - command_started_event:
          command:
            abortTransaction: 1
            lsid: session0
            txnNumber:
            $numberLong: "1"
            startTransaction:
            autocommit: false
            writeConcern:
          command_name: abortTransaction
          database_name: admin
    outcome:
      collection:
        data: []

  - description: abortTransaction does not retry without RetryableWriteError label
    failPoint:
      configureFailPoint: failCommand
      mode: { times: 1 }
      data:
        failCommands: ["abortTransaction"]
        errorCode: 11600 # InterruptedAtShutdown, normally a retryable error code
        errorLabels: []
    operations:
      - name: startTransaction
        object: session0
      - name: insertOne
        object: collection
        arguments:
          session: session0
          document:
            _id: 1
          result:
            insertedId: 1
      - name: abortTransaction
        object: session0
        result:
          errorLabelsOmit: ["RetryableWriteError"]
    expectations:
      - command_started_event:
        command:
          insert: *collection_name
          documents:
            - _id: 1
          ordered: true
          readConcern:
          lsid: session0
          txnNumber:
            $numberLong: "1"
          startTransaction: true
          autocommit: false
          writeConcern:
        command_name: insert
        database_name: *database_name
      - command_started_event:
          command:
            abortTransaction: 1
            lsid: session0
            txnNumber:
            $numberLong: "1"
            startTransaction:
            autocommit: false
            writeConcern:
          command_name: abortTransaction
          database_name: admin
    outcome:
      collection:
          data: []
